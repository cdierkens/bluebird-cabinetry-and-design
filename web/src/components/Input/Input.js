import PropTypes from "prop-types";
import React from "react";

const Input = ({ label, value, placeholder, onChange, id, error, type }) => {
  const className =
    "block w-full bg-white border border-gray-light px-3 py-3 pr-8 leading-tight font-sans text-sm placeholder-gray-darker focus:outline-none focus:shadow-outline";

  return (
    <label className="block relative pb-5 mb-2">
      <span className="block text-blue-dark text-base ml-1 mb-1">{label}</span>

      {type === "text" ? (
        <input
          className={className}
          type="text"
          id={id}
          onChange={onChange}
          placeholder={placeholder}
          value={value}
        />
      ) : (
        <textarea
          className={className}
          id={id}
          onChange={onChange}
          placeholder={placeholder}
          value={value}
          rows={6}
        />
      )}

      {error && (
        <span className="ml-1 text-sm absolute bottom-0 text-gold">
          {error}
        </span>
      )}
    </label>
  );
};

Input.propTypes = {
  label: PropTypes.string.isRequired,
  value: PropTypes.string,
  placeholder: PropTypes.string,
  onChange: PropTypes.func,
  id: PropTypes.string.isRequired,
  error: PropTypes.string,
  type: PropTypes.oneOf(["text", "textarea"]),
};

Input.defaultProps = {
  value: "",
  onChange: () => {},
  placeholder: "",
  error: null,
  type: "text",
};

export default Input;
