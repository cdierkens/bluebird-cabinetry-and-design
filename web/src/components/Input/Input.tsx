import { ErrorMessage, useField } from "formik";
import React from "react";
import { tw } from "../../lib/tailwind";

const className = tw`block w-full bg-white border border-gray-light px-3 py-3 pr-8 leading-tight font-sans text-base placeholder-gray-darker focus:outline-none focus:ring shadow-sm`;

interface InputProps {
  label: string;
  name: string;
  placeholder?: string;
  type?: "text" | "email" | "password" | "textarea";
}

const Input: React.FC<React.PropsWithChildren<InputProps>> = ({ label, type = "text", ...props }) => {
  const [field] = useField({
    ...props,
    type: type === "textarea" ? "text" : type,
  });

  return (
    <label className="block relative pb-5 mb-2">
      <span className="block text-blue-dark text-base ml-1 mb-1">{label}</span>

      {type === "textarea" ? (
        <textarea className={className} rows={6} {...field} {...props} />
      ) : (
        <input className={className} type={type} {...field} {...props} />
      )}

      <ErrorMessage name={field.name}>
        {(error) => (
          <span className="ml-1 text-sm absolute bottom-0 text-gold">
            {error}
          </span>
        )}
      </ErrorMessage>
    </label>
  );
};

export default Input;
