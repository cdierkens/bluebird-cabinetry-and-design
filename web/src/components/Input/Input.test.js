import { render } from "@testing-library/react";
import React from "react";
import Input from "./Input";

it("matches snapshot", () => {
  const { asFragment } = render(<Input label="mock-label" id="mock-id" />);

  expect(asFragment()).toMatchSnapshot();
});
