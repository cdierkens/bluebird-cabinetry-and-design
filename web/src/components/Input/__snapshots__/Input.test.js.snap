// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <label
    class="block relative pb-5 mb-2"
  >
    <span
      class="block text-blue-dark text-base ml-1 mb-1"
    >
      mock-label
    </span>
    <input
      class="block w-full bg-white border border-gray-light px-3 py-3 pr-8 leading-tight font-sans text-sm placeholder-gray-darker focus:outline-none focus:shadow-outline shadow-sm"
      id="mock-id"
      placeholder=""
      type="text"
      value=""
    />
  </label>
</DocumentFragment>
`;
